week3/portfolio_challenge/9_testing_react_with_cypress.md

```diff
@@ -0,0 +1,204 @@
+
+```js
+// cypress/integration/visitor_can_see_landing_page.spec.js
+
+describe('Visitor can see landing page' , () => {
+  it('visitor can see landing page', () => {
+    cy.visit('/')
+  })
+})
+```
+
+To execute cypress functions you need to use `cy` prefix. The most basic one is `visit()` which will navigate to a url which we add. Since we added the baseUlr as `http://localhost:3000` by typing in `visit('/')` we will navigate just there.
+
+After adding that file, start your development server and run cypress and you should see something like this:
+
+![](./cypress_2.png)
+
+We have our first green test! Now let's see what are the most common cypress commands we can use in our tests files to find elements, interact with them and check their content.
+
+
+## Most common cypress commands 
+
+
+### [cy.get()](https://docs.cypress.io/api/commands/get.html#Syntax)
+- Get one or more DOM elements by selector or alias.
+- Examples:
+  - `cy.get(.some-class)` - gets elements with `some-class` class
+  - `cy.get(#some-id)` - gets elements with `some-id` id
+  - `cy.get(button)` - gets all button elements
+  
+### [cy.contains()](https://docs.cypress.io/api/commands/contains.html#Syntax)
+
+- Get the DOM element containing the text. DOM elements can contain more than the desired text and still match.
+- Examples:
+  - `cy.contains('Hello')` - gets elements with `Hello` text
+  - `cy.get('.nav').contains('About')` - gets elements with `About` text **inside** `.nav` class
+
+### [cy.should()](https://docs.cypress.io/api/commands/should.html#Syntax)
+
+- Create an assertion. Assertions are automatically retried until they pass or time out. An alias of [.and()](https://docs.cypress.io/api/commands/and.html#Syntax)
+- Examples:
+  - `cy.get('.error').should('be.empty') ` - Assert that '.error' is empty
+  - `cy.contains('Login').should('be.visible')` - Assert that el is visible
+
+### [cy.click()](https://docs.cypress.io/api/commands/click.html#Syntax)
+
+- Click a DOM element.
+- Examples:
+  - `cy.get('button').click()` - Click on button
+
+### [cy.type()](https://docs.cypress.io/api/commands/type.html#Syntax)
+
+- Type into a DOM element.
+- Examples:
+  - `cy.get('input').type('Hello, World')` - Type 'Hello, World' into the 'input'
+
+
+## Assertions
+
+
+Modify the test file to look like this:
+
+```js
+// cypress/integration/visitor_can_see_landing_page.spec.js
+
+/// <reference types="Cypress" />
+
+describe("Visitor can see landing page", () => {
+  it("visitor can see landing page", () => {
+    cy.visit("/");
+    cy.get("nav")
+      .should("contain", "My Portfolio")
+      .and("contain", "About Me")
+      .and("contain", "My projects");
+    cy.get(".main.container")
+      .should("contain", "Hello World")
+      .and(
+        "contain",
+        "I am studying to become a fullstack junior web developer"
+      );
+  });
+});
+```
```

Comment:
I would like us to be consistent. Single or double quotes?

Reply:
double and done